src/linear_algebra/pi.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2017 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Mario Carneiro, Kevin Buzzard, Yury Kudryashov, Eric Wieser
+-/
+import linear_algebra.basic
+
+/-! ### Pi types of semimodules
+
+This file defines constructors for linear maps whose domains or codomains are pi types.
+
+It contains theorems relating these to each other, as well as to `linear_map.ker`.
+
+## Main definitions
+
+- pi types in the codomain:
+  - `linear_map.proj`
+  - `linear_map.pi`
+
+-/
+
+universes u v w x y z u' v' w' y'
+variables {R : Type u} {K : Type u'} {M : Type v} {V : Type v'} {M₂ : Type w} {V₂ : Type w'}
+variables {M₃ : Type y} {V₃ : Type y'} {M₄ : Type z} {ι : Type x}
+
+namespace linear_map
+
+open function
+open submodule
+
+universe i
+variables [semiring R] [add_comm_monoid M₂] [semimodule R M₂] [add_comm_monoid M₃] [semimodule R M₃]
+{φ : ι → Type i} [∀i, add_comm_monoid (φ i)] [∀i, semimodule R (φ i)]
+
+/-- `pi` construction for linear functions. From a family of linear functions it produces a linear
+function into a family of modules. -/
+def pi (f : Πi, M₂ →ₗ[R] φ i) : M₂ →ₗ[R] (Πi, φ i) :=
+⟨λc i, f i c, λ c d, funext $ λ i, (f i).map_add _ _, λ c d, funext $ λ i, (f i).map_smul _ _⟩
+
+@[simp] lemma pi_apply (f : Πi, M₂ →ₗ[R] φ i) (c : M₂) (i : ι) :
+  pi f c i = f i c := rfl
+
+lemma ker_pi (f : Πi, M₂ →ₗ[R] φ i) : ker (pi f) = (⨅i:ι, ker (f i)) :=
+by ext c; simp [funext_iff]; refl
```

Comment:
```suggestion
by { ext c, simp [funext_iff], refl }
```

Reply:
These proofs are directly copied from the other file - I'm inclined not to touch them just to make the diff easier to follow.

Reply:
Hmm, I think if that's a concern then having the move and style updates in separate commits in the PR branch should be sufficient. But these changes aren't a big deal. Feel free to merge either way.

Reply:
I was planning to do some other further cleanup anyway, so I'll fold these changes in there